pool.go

```diff
@@ -412,11 +408,104 @@ func (p *pool) getNodesToFetch(c cid.Cid, with string) ([]string, error) {
 	}
 
 	// if we still don't have enough nodes, just return the initial set of nodes we got without considering cool off.
-
 	return nodes, nil
 }
 
-func (p *pool) fetchAndUpdate(ctx context.Context, node string, c cid.Cid, attempt int) (blk blocks.Block, err error) {
+func (p *pool) fetchResourceWith(ctx context.Context, path string, cb DataCallback, with string) (err error) {
+	// wait for pool to be initialised
+	<-p.started
+
+	transientErrs := make(map[string]error)
+
+	left := p.config.MaxRetrievalAttempts
+	aff := with
+	if aff == "" {
+		aff = path
+	}
+
+	p.lk.RLock()
+	if left > len(p.endpoints) {
+		left = len(p.endpoints)
+	}
+
+	// if there are no endpoints in the consistent hashing ring, we submit a pool refresh request and fail this fetch.
+	if p.c == nil || p.c.Size() == 0 {
+		p.lk.RUnlock()
+		return ErrNoBackend
+	}
+	nodes, ok := p.c.GetNodes(aff, left)
+	p.lk.RUnlock()
```

Comment:
Please replace the above code with `getNodesToFetch`. You can take a look at how we do it in `fetchBlockWith`